### PR TITLE
[pallet-revive] Eth RPC integration

### PR DESCRIPTION
**Pull request by pgherveou**
This PR introduces the necessary changes to pallet-revive for integrating with our Ethereum JSON-RPC.
The RPC proxy itself will be added in a follow up.

## Changes

-  A new pallet::call  `Call::eth_transact`. This is used as a wrapper to accept unsigned Ethereum transaction, valid call will be routed to `Call::call` or `Call::instantiate_with_code`

- A custom UncheckedExtrinsic struct, that wraps the generic one usually and add the ability to check eth_transact calls sent from an Ethereum JSON-RPC proxy.
- Generated types and traits to support implementing a JSON-RPC Ethereum proxy.

## Flow Overview:
- A user submits a transaction via MetaMask or another Ethereum-compatible wallet.
- The proxy dry run the transaction and add metadata to the call (gas limit in Weight, storage deposit limit, and length of bytecode and constructor input for contract instantiation)
- The raw transaction, along with the additional metadata, is submitted to the node as an unsigned extrinsic.
- On the runtime, our custom UncheckedExtrinsic define a custom Checkable implementation that converts the unsigned extrinsics into checked one
 - It recovers the signer
-  validates the payload, and injects signed extensions, allowing the system to increment the nonce and charge the appropriate fees.
-  re-route the call to pallet-revive::Call::call or pallet-revive::Call::instantiateWithCode

## Dependencies

- https://github.com/koute/polkavm/pull/188

## Follow up PRs
- #5926  
- #6147 (previously #5953)
- #5502

### CODE REVIEW

File: substrate/frame/revive/src/evm/api/rpc_types_gen.rs
Comment (pgherveou) [inline substrate/frame/revive/src/evm/api/rpc_types_gen.rs:1]: The code used to generate rpc_types_gen.rs and rpc_methods_gen.rs will be added in a separate PR to keep the diff manageable here

File: substrate/frame/revive/src/lib.rs [hunk 1]
```diff
@@ -88,7 +91,7 @@ pub use weights::WeightInfo;
 pub use crate::wasm::SyscallDoc;
 
 type TrieId = BoundedVec<u8, ConstU32<128>>;
-type BalanceOf<T> =
+pub type BalanceOf<T> =
```
Comment (athei) [inline substrate/frame/revive/src/lib.rs:94]: I don't think this is used anywhere? Even if so I don't think we should export those short hand aliases.

File: substrate/frame/revive/src/evm/runtime.rs [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+{
+	fn from(
+		utx: generic::UncheckedExtrinsic<MultiAddress, Call, MultiSignature, E::Extra>,
+	) -> Self {
+		Self(utx)
+	}
+}
+
+impl<Call, E: EthExtra>
+	Into<generic::UncheckedExtrinsic<MultiAddress, Call, MultiSignature, E::Extra>>
```
Comment (athei) [inline substrate/frame/revive/src/evm/runtime.rs:81]: impl `From` instead

File: substrate/bin/node/runtime/src/lib.rs [hunk 3]
```diff
@@ -2560,11 +2572,37 @@ pub type SignedExtra = (
...
+			frame_metadata_hash_extension::CheckMetadataHash::<Runtime>::new(false),
+			pallet_revive::evm::runtime::CheckEthTransact::<Runtime>::from(eth_fee),
+		)
+	}
+}
+
 /// Unchecked extrinsic type as expected by this runtime.
 pub type UncheckedExtrinsic =
-	generic::UncheckedExtrinsic<Address, RuntimeCall, Signature, SignedExtra>;
+	pallet_revive::evm::runtime::UncheckedExtrinsic<RuntimeCall, EthSignedExtra>;
```
Comment (athei) [inline substrate/bin/node/runtime/src/lib.rs:2604]: I don't get this. Shouldn't this break all wallets since you replace `SignedExtra` (type that contains data) with `EthSignedExtra` (zero sized type) for all extrinsics?

Comment (athei) [reply substrate/bin/node/runtime/src/lib.rs:2604]: It actually uses `EthSignedExtra::Extra` and hence leaves the on wrote format of the extrinsic untouched.

File: substrate/frame/revive/src/evm/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+	/// The gas fee, computed from the signed Ethereum transaction.
+	///
+	/// # Note
+	///
+	/// This is marked as `#[codec(skip)]` as this extracted from the Ethereum transaction and
+	/// should have a zero size encoded representation.
+	#[codec(skip)]
+	eth_fee: Option<BalanceOf<T>>,
+	_phantom: core::marker::PhantomData<T>,
+}
```
Comment (athei) [inline substrate/frame/revive/src/evm/runtime.rs:209]: It causes quite the code churn to add this fake signed extension everwhere. Why not to the `eth_fee` check directly in `Checkable::check`?

Comment (pgherveou) [reply substrate/frame/revive/src/evm/runtime.rs:209]: I did not have that initially, but then realized I needed the encoded_len of the extrinsic to properly estimate the fees. We could use [encoded_size](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/trait.Encode.html#method.encoded_size) to recalculate it, but not sure this is much better 


>  Its also super confusing to add a SignedExtension specifically for an unsigned transaction

SignedExtension can be used for all transaction (all methods on the trait have _unsigned variant). 

Relevant code running this logic:
https://github.com/paritytech/polkadot-sdk/blob/pg/revive-rpc-server/substrate/frame/executive/src/lib.rs?plain=1#L329-L349

Comment (athei) [reply substrate/frame/revive/src/evm/runtime.rs:205]: > I did not have that initially, but then realized I needed the encoded_len of the extrinsic to properly estimate the fees. We could use [encoded_size](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/trait.Encode.html#method.encoded_size) to recalculate it, but not sure this is much better

I think it would make the code much easier to understand in my mind.

Comment (athei) [reply substrate/frame/revive/src/evm/runtime.rs:205]: I think it would make the code easier to understand. Not having this additional signed extension.

Comment (pgherveou) [reply substrate/frame/revive/src/evm/runtime.rs:204]: Ok I will need to move a bunch of constraint but I'll give it a shot,
It just seems a bit wasteful to get get the dispatch_info and recompute the length here, when all of this is provided to the signed_extension later on.

Comment (pgherveou) [reply substrate/frame/revive/src/evm/runtime.rs]: done the changes here https://github.com/paritytech/polkadot-sdk/pull/5866/commits/222a557ef22dceb9c3d8499c20cd0406e6d61645

File: substrate/frame/revive/src/exec.rs [hunk 5]
```diff
@@ -1020,6 +1033,7 @@ where
 					frame.contract_info.get(&frame.account_id),
 					executable.code_info(),
 				)?;
+
```
Comment (pgherveou) [inline substrate/frame/revive/src/exec.rs:1036]:
```suggestion
```

File: substrate/frame/revive/src/lib.rs [hunk 6]
```diff
@@ -1212,7 +1390,20 @@ sp_api::decl_runtime_apis! {
...
+		///
+		/// See [`crate::Pallet::bare_eth_transact`]
+		fn eth_transact(
+			origin: H160,
+			dest: Option<H160>,
+			value: Balance,
+			input: Vec<u8>,
+			gas_limit: Option<Weight>,
+			storage_deposit_limit: Option<Balance>,
+		) -> EthContractResult<Balance>;
```
Comment (athei) [inline substrate/frame/revive/src/lib.rs:1406]: Why do we need that? Can't the proxy dispatch into the existing runtime APIs?

File: substrate/frame/revive/src/evm/runtime.rs [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+use alloc::{vec, vec::Vec};
+
+/// Some way of identifying an account on the chain.
+pub type AccountId = AccountId32;
+
+/// The type for looking up accounts. We don't expect more than 4 billion of them.
+pub type AccountIndex = u32;
+
+/// The address format for describing accounts.
+pub type MultiAddress = sp_runtime::MultiAddress<AccountId, AccountIndex>;
```
Comment (athei) [inline substrate/frame/revive/src/evm/runtime.rs:55]: Do we need to hardcode these or can we just have them as type parameters?

File: substrate/frame/revive/src/evm/runtime.rs [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+			}
+		};
+
+		let nonce = nonce.try_into().map_err(|_| InvalidTransaction::Call)?;
+		let eth_fee =
+			gas_price.saturating_mul(gas).try_into().map_err(|_| InvalidTransaction::Call)?;
+		let info = call.get_dispatch_info();
+		let function: Call = call.into();
+
+		let tip = Default::default();
```
Comment (athei) [inline substrate/frame/revive/src/evm/runtime.rs:312]: When `eth_fee` is larger than `actual_fee` we probably should supply the difference as `tip`. This would enable people to prioritise their transaction by supplying a higher gas fee. Just as they would on eth. Not sure how they determine the `gas_price` in the first place, though.

File: substrate/bin/node/runtime/src/lib.rs [hunk 9]
```diff
@@ -2532,6 +2533,16 @@ mod runtime {
...
+impl TryFrom<RuntimeCall> for pallet_revive::Call<Runtime> {
+	type Error = ();
+
+	fn try_from(value: RuntimeCall) -> Result<Self, Self::Error> {
+		match value {
+			RuntimeCall::Revive(call) => Ok(call),
+			_ => Err(()),
+		}
+	}
+}
```
Comment (athei) [inline substrate/bin/node/runtime/src/lib.rs:2545]: Where is this used?

File: substrate/frame/revive/src/primitives.rs [hunk 10]
```diff
@@ -76,19 +76,63 @@ pub struct ContractResult<R, Balance, EventRecord> {
...
+	) -> EthContractResult<Balance> {
+		EthContractResult {
+			transact_kind: self.transact_kind,
+			result: self.result,
+			gas_limit: self.gas_limit,
+			storage_deposit: self.storage_deposit,
+			fee: compute_fee(self.len, &self.dispatch_info, 0.into()),
+		}
+	}
+}
```
Comment (athei) [inline substrate/frame/revive/src/primitives.rs:120]: I don't think we need this. We are already tightly coupled with `pallet-transaction-payment` and can just calculate the fee inside `bare_eth_transact`.  No need for the details type.

File: substrate/bin/node/runtime/src/lib.rs (hunk 9 above)
Comment (pgherveou) [reply substrate/bin/node/runtime/src/lib.rs:2545]: here https://github.com/paritytech/polkadot-sdk/blob/pg/revive-rpc-fixes-v2/substrate/frame/revive/src/evm/runtime.rs?plain=1#L135

File: substrate/frame/revive/src/tests.rs [hunk 11]
```diff
@@ -509,6 +523,17 @@ impl Config for Test {
...
+impl TryFrom<RuntimeCall> for crate::Call<Test> {
+	type Error = ();
+
+	fn try_from(value: RuntimeCall) -> Result<Self, Self::Error> {
+		match value {
+			RuntimeCall::Contracts(call) => Ok(call),
+			_ => Err(()),
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/src/tests.rs:535]: For the checkable implementation.
We can probably add some comment

https://github.com/paritytech/polkadot-sdk/blob/4be191641ca47e0c461adfa2dd8159834256ff89/substrate/frame/revive/src/evm/runtime.rs?plain=1#L133-L135

File: substrate/frame/revive/src/lib.rs [hunk 12]
```diff
@@ -738,6 +740,41 @@ pub mod pallet {
...
+				EthTransactKind::InstantiateWithCode{code_len, data_len} => T::WeightInfo::instantiate_with_code(*code_len, *data_len)
+					.saturating_add(*gas_limit)
+			}
+		)]
+		pub fn eth_transact(
+			origin: OriginFor<T>,
+			payload: Vec<u8>,
+			gas_limit: Weight,
+			#[pallet::compact] storage_deposit_limit: BalanceOf<T>,
+			transact_kind: EthTransactKind,
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/src/lib.rs:773]: ha right might be able to simplify that now that we have the blob_length from polkavm

[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/src/lib.rs:773]: Yeah not sure why I added that now getting rid of it

File: substrate/frame/revive/src/lib.rs (hunk 6 above)
Comment (pgherveou) [reply substrate/frame/revive/src/lib.rs:1382]: this let us do the branching logic here between the two type of transaction, plus we compute the exact fee here,

File: substrate/frame/revive/src/evm/runtime.rs (hunk 8 above)
Comment (pgherveou) [reply substrate/frame/revive/src/evm/runtime.rs:290]: we still want to ensure that the injected weight match the fees the user specified in the transaction.
Since we hard code the gas_price to 1,  we can do our check using this value, and if the user provided a gas_price greater than 1,  use the difference as tip.

for post eip1559 tx types there is a maxPriorityFeePerGas that define the tip value

will amend to add the tip calculation.

File: substrate/frame/revive/src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/src/lib.rs]: ah with that change, to get the exact fee estimate, I need to build the eth_transact dispatch info to get the encoded length but also the actual dispatch_info from call or instantiate to compute the exact fee, fixing my changes now

File: substrate/bin/node/runtime/src/lib.rs (hunk 9 above)
Comment (athei) [reply substrate/bin/node/runtime/src/lib.rs:2545]: Got it thanks

File: substrate/frame/revive/src/evm/runtime.rs [hunk 13]
```diff
@@ -0,0 +1,636 @@
...
+	RuntimeDebug, Saturating,
+};
+
+use alloc::vec::Vec;
+
+/// The EVM gas price.
+/// We use a fixed value for the gas price.
+/// This let us calculate the gas estimate for a transaction with the formula:
+/// `estimate_gas = substrate_fee / gas_price`.
+pub const GAS_PRICE: u128 = 1_000u128;
```
Comment (athei) [inline substrate/frame/revive/src/evm/runtime.rs:47]: Need some more explanation what "we use" means here. I assume that means this is what we return from some RPC and this is also why this is public (it is used in the proxy).

Also why don't we use `1`?

File: substrate/frame/revive/src/evm/runtime.rs [hunk 14]
```diff
@@ -0,0 +1,636 @@
...
+		let eth_fee: BalanceOf<Self::Config> = U256::from(gas_price)
+			.saturating_mul(gas)
+			.try_into()
+			.map_err(|_| InvalidTransaction::Call)?;
+
+		let info = call.get_dispatch_info();
+		let function: Call = call.into();
+
+		// Fees calculated from the extrinsic, without the tip.
+		let actual_fee_no_tip: BalanceOf<Self::Config> =
```
Comment (athei) [inline substrate/frame/revive/src/evm/runtime.rs:311]: I think its confusing to have the `no_tip` attached even though its always without tip.

```suggestion
		let actual_fee: BalanceOf<Self::Config> =
```

File: substrate/frame/revive/src/evm/runtime.rs (hunk 13 above)
Comment (pgherveou) [reply substrate/frame/revive/src/evm/runtime.rs]: yes this is shared with the proxy, I moved it from the proxy to here so we can share the constant.
I was using 1 but then figured that this gas_price that we return from the rpc might be used by the wallet to adjust how much tip it wants to give to the tx. Wallet live in a world where they can use fractional number so 1 might still be fine.

File: substrate/frame/revive/src/evm/runtime.rs [hunk 15]
```diff
@@ -0,0 +1,636 @@
...
+impl<Address, Call, Signature, E> GetDispatchInfo
+	for UncheckedExtrinsic<Address, Call, Signature, E>
+where
+	Call: GetDispatchInfo,
+	E: EthExtra,
+{
+	fn get_dispatch_info(&self) -> DispatchInfo {
+		self.0.get_dispatch_info()
+	}
+}
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/src/evm/runtime.rs:171]: no we want it like this so that when you ask the weight of eth_transact it returns MAX
it's only used by runtime api api like  tx_payment to get query_fee_details as far as I can tell

File: substrate/bin/node/runtime/src/lib.rs [hunk 16]
```diff
@@ -3006,14 +3045,38 @@ impl_runtime_apis! {
...
+			dest: Option<H160>,
+			value: Balance,
+			input: Vec<u8>,
+			gas_limit: Option<Weight>,
+			storage_deposit_limit: Option<Balance>,
+		) -> pallet_revive::EthContractResult<Balance>
+		{
+			use pallet_revive::AddressMapper;
+			let blockweights: BlockWeights = <Runtime as frame_system::Config>::BlockWeights::get();
+			let origin = <Runtime as pallet_revive::Config>::AddressMapper::to_account_id_contract(&from);
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/bin/node/runtime/src/lib.rs:3059]: yeah good point

File: substrate/frame/revive/fixtures/contracts/demo.rs
Comment (athei) [inline substrate/frame/revive/fixtures/contracts/demo.rs:1]: Lefover from the merge mishap?

File: substrate/frame/revive/src/evm/runtime.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (athei) [reply substrate/frame/revive/src/evm/runtime.rs:171]: Isn't it also used during block building to quickly estimate if the tx will fit? Seems dangerous to me to return the wrong thing here.

File: substrate/frame/revive/src/evm/runtime.rs (hunk 13 above)
Comment (athei) [reply substrate/frame/revive/src/evm/runtime.rs]: Yes but can you put that in the doc that this constant is used by the proxy to advertise it via the gas price rpc?

File: substrate/bin/node/runtime/src/lib.rs (hunk 9 above)
Comment (pgherveou) [reply substrate/bin/node/runtime/src/lib.rs:2601]: yea we could update the proc macro to generate it for us

File: substrate/frame/revive/src/evm/runtime.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/src/evm/runtime.rs]: It's not the executive do it like this
https://github.com/paritytech/polkadot-sdk/blob/483b6385eb631d412e0124f6169256768ac5a9cc/substrate/frame/executive/src/lib.rs?plain=1#L785-L787

it's really just these 2 runtime APIs query_fee_details.

> Seems dangerous to me to return the wrong thing here.

Well we return the truth :), would be pretty bad that the logic to get the underlying call and it's dispatch info was executed twice.

[the remark this answers is not in the document]
Comment (athei) [reply substrate/frame/revive/src/evm/runtime.rs]: Its also called in `validate_transaction` but only on checked transaction. So this should be fine:
https://github.com/paritytech/polkadot-sdk/blob/483b6385eb631d412e0124f6169256768ac5a9cc/substrate/frame/executive/src/lib.rs?plain=1#L868-L874

I am wondering if the unchecked transaction should even imlement `GetDispatchInfo`. It seems like a good gun. Those `query_fee_details` runtime APIs will return the wrong values.

**Review by athei — APPROVED** [review]

**Review by xermicus — APPROVED** [review]

File: substrate/bin/node/runtime/Cargo.toml [hunk 17]
```diff
@@ -56,6 +57,8 @@ std = [
 	"primitive-types/std",
 	"scale-info/std",
 	"serde_json/std",
+	"sp-debug-derive/std",
+	"sp-debug-derive/std",
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/bin/node/runtime/Cargo.toml:61]: ahah just to be safe add it multiple time :)
